src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/UnsignedAdapters.java

```diff
@@ -0,0 +1,159 @@
+            return intToUnsignedByte(target);
+        } else if (adaptedType == int.class && carrier == short.class) {
+            return intToUnsignedShort(target);
+        } else if (adaptedType == long.class && carrier == byte.class) {
+            return longToUnsignedByte(target);
+        } else if (adaptedType == long.class && carrier == short.class) {
+            return longToUnsignedShort(target);
+        } else if (adaptedType == long.class && carrier == int.class) {
+            return longToUnsignedInt(target);
+        } else {
+            throw new InternalError("should not reach here");
+        }
+    }
+
+    // int to byte
+    private static VarHandle intToUnsignedByte(VarHandle target) {
+        if (target.varType() != byte.class)
+            throw new InternalError("expected byte carrier, but got: " + target.varType());
+        return MemoryHandles.filterValue(target, INT_TO_BYTE, BYTE_TO_UNSIGNED_INT);
+    }
+
+    // int to short
+    private static VarHandle intToUnsignedShort(VarHandle target) {
+        if (target.varType() != short.class)
+            throw new InternalError("expected byte carrier, but got: " + target.varType());
+        return MemoryHandles.filterValue(target, INT_TO_SHORT, SHORT_TO_UNSIGNED_INT);
+    }
+
+    // long to byte
+    private static VarHandle longToUnsignedByte(VarHandle target) {
+        if (target.varType() != byte.class)
+            throw new InternalError("expected byte carrier, but got: " + target.varType());
+        return MemoryHandles.filterValue(target, LONG_TO_BYTE, BYTE_TO_UNSIGNED_LONG);
+    }
+
+    // long to short
+    private static VarHandle longToUnsignedShort(VarHandle target) {
+        if (target.varType() != short.class)
+            throw new InternalError("expected byte carrier, but got: " + target.varType());
+        return MemoryHandles.filterValue(target, LONG_TO_SHORT, SHORT_TO_UNSIGNED_LONG);
+    }
+
+    //long to int
+    private static VarHandle longToUnsignedInt(VarHandle target) {
+        if (target.varType() != int.class)
+            throw new InternalError("expected byte carrier, but got: " + target.varType());
+        return MemoryHandles.filterValue(target, LONG_TO_INT, INT_TO_UNSIGNED_LONG);
+    }
+
+    private static void checkWidenable(Class<?> carrier) {
+        if (!(carrier == byte.class || carrier == short.class || carrier == int.class)) {
+            throw newIllegalArgumentException("illegal carrier", carrier.getSimpleName());
+        }
+    }
+
+    private static void checkNarrowable(Class<?> type) {
+        if (!(type == int.class || type == long.class)) {
+            throw newIllegalArgumentException("illegal adapter type", type.getSimpleName());
+        }
+    }
+
+    private static void checkTargetWiderThanCarrier(Class<?> carrier, Class<?> target) {
+        if (Wrapper.forPrimitiveType(target).bitWidth() <= Wrapper.forPrimitiveType(carrier).bitWidth()) {
```

Comment:
Note that MemoryHandles has already a routine to retrieve carrier size - maybe there's some reuse possible there